pyanaconda/core/glib.py

```diff
@@ -46,16 +50,121 @@
            "MAXUINT"]
 
 
-def create_main_loop():
+def create_main_loop(main_context=None):
     """Create GLib main loop.
 
+    :param main_context: main context to be used for the loop
+    :type main_context: GLib.MainContext
     :returns: GLib.MainLoop instance.
     """
-    return MainLoop()
+    return MainLoop(main_context)
 
 
 def create_new_context():
     """Create GLib context.
 
     :returns: GLib.MainContext."""
     return MainContext.new()
+
+
+def error_is_cancelled(e):
+    """Is the error indicating that operation was cancelled?
+
+    :returns: True if operation was cancelled, False otherwise
+    :rtype: bool
+    """
+    if isinstance(e, GError):
+        if e.domain == "g-io-error-quark" and e.code == IOErrorEnum.CANCELLED:
+            return True
+    return False
+
+
+class MainLoopRun:
```

Comment:
The async_call_finish method returns tuple in one case and single value in other. That was probably the main thing why I didn't want to go further abstracting. At least until we have more than 2 instances to abstract (I realize it is my - perhaps not correct - internal rule - to have more than 2 instances:)).
And that is why context manager looked better to me - there is quite some content that is instance specific.
But perhaps it can be unified anyway, I'd just not do it now.

Reply:
The tuple is not a problem. It doesn't matter if the return value is just one value or a list or a tuple or whatever, you can deal with it later when you process the result of the call (together with errors and timeouts). You know what you called, so you know what is the type of the return value. You can use `res, con = result.received_data` to handle the result of `add_connection2` and `res = result.received_data` to handle the result of `update2`.